Introduce a new Products by Tag(s) block

mikejolley opened the pull request:
**NOTE: I'm targeting the abstract block refactor branch so I can take advantage of those changes. This will need rebasing with master before merge once the other PR is in.

This PR introduces a products by tag block (based heavily on the products by categories/attributes blocks that exist already).

To prevent scaling issues, if there are large numbers of tags, rather than list all (like the category widget) it will only list all if there are < 20 tags. If there are more, 20 most popular tags will be shown until you search. The search is debounced and based on the code in https://github.com/woocommerce/woocommerce-gutenberg-products-block/pull/547

Fixes #503 

### Screenshots

![Settings](https://user-images.githubusercontent.com/90977/57367135-54903400-7180-11e9-8171-c8dcf35357d4.png)

### How to test the changes in this Pull Request:

1. Checkout/build branch
2. Ensure you have tags in your catalog (good to test < 20 and > 20 since the search behaviour will be different)
3. Add the "Products by Tag" block to a post and check the options/output.

mikejolley commented:
> Can you break this apart into smaller PRs? Split apart the new core code, API work, the new control, and the block? You can chain them to make each build upon the last, but splitting them apart will make it easier to review.

I could do 1 for the API/php work and one for the block if needed but since most of the code is in new files it probably won't be much easier to review. If I touched a bunch of existing files I'd understand why splitting this up was necessary :) I see past PRs where blocks and controls were added were not split (https://github.com/woocommerce/woocommerce-gutenberg-products-block/pull/276/).

> Also, how did you come up with 20 as the tag limit?

How did you come up with the 200 product limit in #547? WP core screens tend to paginate around 20, so I thought it would make sense to limit the first request to that but I'm not tied to it. Do you have a better number in mind? 

Let me know about the limit and Ill split this into 2.

ryelle commented:
Sure, 2 PRs is fine.

> How did you come up with the 200 product limit in #547?

Well, the initial idea was unlimited so we could take advantage of faster searching (in JS, as opposed to making new API calls with each search). We can make API calls in batches of 100, so I looked at the average size of stores (using JP data), and <200 would cover almost 95% of stores and only require at most 2 API calls. Larger sites then use the more API-heavy search, to prevent the long initial load mentioned in #505.

So I would look at the load time of the API at different limits, and maybe see if you can get any data about average number of tags on stores.

timmyc commented:
Regarding the number of tags - should we just pick a number between 20 and 100 and see how it performs for now? We could do a data request to see some averages/means for number of product tags too and revisit after we have those results.

mikejolley commented:
@ryelle I'm going to mark these as blocked since we need dynamic blocks and API changes first.

Aljullu commented on webpack.config.js:
```diff
@@ -38,6 +38,7 @@ const GutenbergBlocksConfig = {
 		'product-top-rated': './assets/js/blocks/product-top-rated/index.js',
 		'products-by-attribute': './assets/js/blocks/products-by-attribute/index.js',
 		'featured-product': './assets/js/blocks/featured-product/index.js',
+		'product-tag': './assets/js/blocks/product-tag/index.js',
```
Can be done in a follow-up (or we can create a `good first issue` ticket), but this block should probably be renamed to `product-by-tag` in order to keep it consistent with other blocks.

mikejolley replied:
There is a `product-category` which this block follows. There is already inconsistency :(

Aljullu replied:
Right, and I see it's not as easy to fix as I thought, because once the blocks have been added to a post, we can't change their name without breaking them. Let's leave it like that for now, then.

Aljullu commented on tests/bootstrap.php:
```diff
@@ -25,7 +25,7 @@ function() {
 		WC_Install::install();
 		$GLOBALS['wp_roles'] = null; // WPCS: override ok.
-		wp_roles();
+			wp_roles();
```
Indentation issue:
```suggestion
		wp_roles();
```

Aljullu commented on tests/bootstrap.php:
```diff
@@ -2,7 +2,7 @@
  * PHPUnit bootstrap file
  *
- * @package WooCommerce\Blocks
+ * @package Sample_Plugin
```
Is that a mistake?

Aljullu commented on src/BlockTypes/ProductTag.php:
```diff
@@ -0,0 +1,59 @@
+	protected $block_name = 'product-tag';
+
+	/**
+	 * This function is not necessary in this block.
```
I don't understand this comment. :confused:

mikejolley replied:
Copied from product category :p Ill improve both.

Aljullu commented on assets/js/components/utils/index.js:
```diff
@@ -43,3 +44,38 @@ export const getProducts = ( { selected = [], search } ) => {
+	const requests = [
+		addQueryArgs( '/wc/v3/products/tags', {
+			per_page: limitTags ? 100 : -1,
+			orderby: limitTags ? 'count' : 'name',
```
Out of curiosity, why are we changing the _orderby_ depending on this? Was that in the specs or is there another reason for that?

mikejolley replied:
I did this so that if we're only showing a handful of tags they are name sorted (so you can find the tag you want easily or search), but if you have pages and pages of tags you get the tags with the most products first, not empty or less-used tags.  Basically, most used first.

Aljullu commented on assets/js/blocks/product-tag/editor.scss:
```diff
@@ -0,0 +1,9 @@
+.wc-block-product-tag__selection {
+	width: 100%;
+}
+.wc-block-product-tag__cancel-button.is-tertiary {
```
This selector doesn't seem to match any button. I guess we might need to change the class in `block.js` or change the selector here.

Aljullu commented on assets/js/components/product-tag-control/style.scss:
```diff
@@ -0,0 +1,16 @@
+.woocommerce-product-tags__operator {
```
All styles in this files are only shown in the editor, right? Should they be moved to `editor.scss`?

mikejolley replied:
No, this is a component not a block. Other components use `style.scss`. Only the blocks use `editor.scss`

Aljullu replied:
Which other components?

`style.scss` will be included in the store frontend, so my understanding is that it shouldn't contain any styles from the admin, right? Otherwise we are polluting store's styles with unnecessary CSS rules. :worried:

mikejolley commented:
Thanks for the review @Aljullu. I've added the spinner and fixed those other mistakes. 

I also noticed there were no placeholders when no tags were added so I added those too:

![Edit Post ‹ local wordpress test — WordPress 2019-07-09 12-27-14](https://user-images.githubusercontent.com/90977/60884575-204ef780-a245-11e9-898e-e1ee58cf43ff.png)

And I also added a blocks endpoint for tags to avoid using v3 directly. This should be ready to re-review now.

mikejolley commented:
> Not a big deal, but something we might want to improve in the future.

@Aljullu Agreed. I also don't like that selected items are hidden when the spinner is on, but this is the SearchListComponent in wc-admin.
